pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionAssignmentTailer.java

```diff
@@ -145,7 +145,7 @@ private Thread getTailerThread() {
                 try {
                     Message<byte[]> msg = reader.readNext(1, TimeUnit.SECONDS);
                     if (msg == null) {
-                        if (exitOnEndOfTopic && !reader.hasMessageAvailable()) {
+                        if (exitOnEndOfTopic && !reader.hasMessageAvailableAsync().get(10, TimeUnit.SECONDS)) {
```

Comment:
It's dangerous to have a timeout here, because we might miss messages. Instead we should rely on the timeout in the `GetLastMessageId` RPC (which should be already there now)

Reply:
When the timeout expires, a TimeoutException will be thrown and which will force the worker to shutdown and restart.  Another worker can more process to becoming leader.